dataframe/dataframe.go

```diff
@@ -400,13 +463,317 @@ func (df *DataFrame) stringify() string {
 	return answer
 }
 
-func dtypeFromWhich(which int) string {
-	if which == typeInt {
-		return "int64"
-	} else if which == typeFloat {
-		return "float64"
+// apply method iterates the rows of a DataFrame, calls the given function for
+// each row, creating a single Series of the results
+func dataframeApply(thread *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		funcVal, axisVal starlark.Value
+	)
+	self := b.Receiver().(*DataFrame)
+
+	if err := starlark.UnpackArgs("apply", args, kwargs,
+		"function", &funcVal,
+		"axis?", &axisVal,
+	); err != nil {
+		return nil, err
+	}
+
+	axis, ok := toIntMaybe(axisVal)
+	if !ok || axis != 1 {
+		return nil, fmt.Errorf("axis must equal 1 for row-wise application")
+	}
+
+	funcObj, ok := funcVal.(*starlark.Function)
+	if !ok {
+		return nil, fmt.Errorf("first argument must be a function")
+	}
+
+	var result []string
+	for rows := newRowIter(self); !rows.Done(); rows.Next() {
+		r := rows.GetRow()
+		arguments := r.toTuple()
+		res, err := starlark.Call(thread, funcObj, arguments, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		text, ok := res.(starlark.String)
+		if !ok {
+			return nil, fmt.Errorf("fn.apply should have returned String")
+		}
```

Comment:
only accepting string return types is pretty limited. In the future will we coerce non-string return values to a string if it's non-string, keeping `Series.valObjs` as type `[]string`, or convert the field to a generic type & add complexity to `Series`?

serializing to & from strings is expensive, which has me leaning toward the latter

Reply:
No, it's just a hack for now. It was also written long ago before I had `typedSliceBuilder`, which is able to do that sort of type wrangling with ease. I'll add a TODO to fix this.